drivers/serial/uart_sam0.c

```diff
@@ -0,0 +1,292 @@
+
+static void uart_sam0_irq_callback_set(struct device *dev,
+				       uart_irq_callback_t cb)
+{
+	struct uart_sam0_dev_data *const dev_data = DEV_DATA(dev);
+
+	dev_data->cb = cb;
+}
+#endif
+
+static const struct uart_driver_api uart_sam0_driver_api = {
+	.poll_in = uart_sam0_poll_in,
+	.poll_out = uart_sam0_poll_out,
+#if CONFIG_UART_INTERRUPT_DRIVEN
+	.fifo_read = uart_sam0_fifo_read,
+	.irq_tx_disable = uart_sam0_irq_tx_disable,
+	.irq_rx_enable = uart_sam0_irq_rx_enable,
+	.irq_rx_disable = uart_sam0_irq_rx_disable,
+	.irq_rx_ready = uart_sam0_irq_rx_ready,
+	.irq_is_pending = uart_sam0_irq_is_pending,
+	.irq_update = uart_sam0_irq_update,
+	.irq_callback_set = uart_sam0_irq_callback_set,
+#endif
+};
+
+#if CONFIG_UART_INTERRUPT_DRIVEN
+#define UART_SAM0_INIT_DEFN(addr, n)					       \
+	static int uart_sam0_init_##n(struct device *dev)		       \
+	{								       \
+		IRQ_CONNECT(CONFIG_UART_SAM0_SERCOM##n##_IRQ,		       \
+			    CONFIG_UART_SAM0_SERCOM##n##_IRQ_PRIORITY,	       \
+			    uart_sam0_isr, DEVICE_GET(uart_sam0_##n), 0);      \
+		irq_enable(CONFIG_UART_SAM0_SERCOM##n##_IRQ);		       \
+		return uart_sam0_init(dev);				       \
+	}
+#else
+#define UART_SAM0_INIT_DEFN(n)						       \
+	static int uart_sam0_init_##n(struct device *dev)		       \
+	{								       \
+		return uart_sam0_init(dev);				       \
+	}
+#endif
+
+#define UART_SAM0_CONFIG_DECL(n)					       \
+static const struct uart_sam0_dev_cfg uart_sam0_config_##n = {		       \
+	.regs = (SercomUsart *)CONFIG_UART_SAM0_SERCOM##n##_BASE_ADDRESS,      \
+	.baudrate = CONFIG_UART_SAM0_SERCOM##n##_CURRENT_SPEED,		       \
+	.pm_apbcmask = PM_APBCMASK_SERCOM##n,				       \
+	.gclk_clkctrl_id = GCLK_CLKCTRL_ID_SERCOM##n##_CORE,		       \
+	.ctrla = SERCOM_USART_CTRLA_RXPO(3) | SERCOM_USART_CTRLA_TXPO(1),      \
+	.pin_rx = PIN_UART_SAM0_SERCOM##n##_RX,				       \
+	.pin_tx = PIN_UART_SAM0_SERCOM##n##_TX,				       \
+}
+
+#define UART_SAM0_DEVICE_INIT(n)					       \
+	static int uart_sam0_init_##n(struct device *dev);		       \
+									       \
+	static struct uart_sam0_dev_data uart_sam0_data_##n = {};	       \
+	UART_SAM0_CONFIG_DECL(n);                                              \
+									       \
+	DEVICE_AND_API_INIT(uart_sam0_##n, CONFIG_UART_SAM0_SERCOM##n##_LABEL, \
+			    uart_sam0_init_##n, &uart_sam0_data_##n,	       \
+			    &uart_sam0_config_##n, PRE_KERNEL_1,	       \
+			    CONFIG_KERNEL_INIT_PRIORITY_DEVICE,		       \
+			    &uart_sam0_driver_api);			       \
+	UART_SAM0_INIT_DEFN(addr, n)
+
+#if CONFIG_UART_SAM0_SERCOM0_BASE_ADDRESS
+UART_SAM0_DEVICE_INIT(0)
+#endif
+
+#if CONFIG_UART_SAM0_SERCOM1_BASE_ADDRESS
+UART_SAM0_DEVICE_INIT(1)
+#endif
+
+#if CONFIG_UART_SAM0_SERCOM2_BASE_ADDRESS
+UART_SAM0_DEVICE_INIT(2)
+#endif
+
+#if CONFIG_UART_SAM0_SERCOM3_BASE_ADDRESS
+UART_SAM0_DEVICE_INIT(3)
+#endif
+
+#if CONFIG_UART_SAM0_SERCOM4_BASE_ADDRESS
+UART_SAM0_DEVICE_INIT(4)
+#endif
+
+/* SERCOM5 on RX=PB23, TX=PB22 */
```

Comment:
We should remove the comment, alternatively keep the SERCOM5 and remove pin names. In this case we would need to add comments for SERCOM0-SERCOM4.

Reply:
Oops, those are now in soc_pinmap.h.  Fixed.